### PR TITLE
refactor(pubsub): split SubscriptionSession [1]

### PR DESCRIPTION
**Pull request by coryan**
Copy the functionality to perform message lease management to a separate
class. For now, this class is only used in tests. A future PR will
replace `pubsub_internal::SubscriptionSession` with a class that uses
this new component.

Part of the work for #4970 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4990)

### CODE REVIEW

File: google/cloud/pubsub/internal/subscription_batch_source.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * A message source generates messages via `BatchCallback` callbacks. Typically
+ * messages are obtained by calling `AsyncPull()` on a `SubscriberStub`, but we
+ * also need to mock this class in our tests and we implement message lease
+ * management as a decorator.
+ */
+class SubscriptionBatchSource {
+ public:
+  virtual ~SubscriptionBatchSource() = default;
+
+  /// Shutdown the source, cancel any outstanding requests and or timers, no
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_batch_source.h:44]: s/, no/. No/

File: google/cloud/pubsub/internal/subscription_batch_source.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  /// callbacks should be generated after this call.
+  virtual void Shutdown() = 0;
+
+  /**
+   * Positive acknowledgement the message associated with @p ack_id.
+   *
+   * The application has successfully handled this message and no new deliveries
+   * are necessary. The @p size parameter should be the original message size
+   * estimate, it may be used by the message source to flow control large
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_batch_source.h:53]: s/, it/. It/

File: google/cloud/pubsub/internal/subscription_batch_source.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+   * messages.
+   */
+  virtual future<Status> AckMessage(std::string const& ack_id,
+                                    std::size_t size) = 0;
+
+  /**
+   * Negative acknowledgement for message associated with @p ack_id.
+   *
+   * The application has not able to handle this message, the service should
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_batch_source.h:62]: s/, the/, so the/ perhaps?

File: google/cloud/pubsub/internal/subscription_batch_source.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+   * re-deliver the message (subject to the topic and subscription
+   * configuration). The @p size parameter should be the original message size
+   * estimate, it may be used by the message source to flow control large
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_batch_source.h:65]: s/, it/. It/

File: google/cloud/pubsub/internal/subscription_lease_management_test.cc [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    m.set_ack_id("ack-" + id);
+    m.mutable_message()->set_message_id("message-" + id);
+  }
+  return response;
+}
+
+TEST(SubscriptionLeaseManagementTest, NormalLifecycle) {
+  auto stub = std::make_shared<pubsub_testing::MockSubscriberStub>();
+
+  auto constexpr kTestDeadlineSeconds = 345;
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_lease_management_test.cc:46]: Consider making these `std::chrono::seconds`.

File: google/cloud/pubsub/internal/subscription_lease_management_test.cc [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    return acks;
+  };
+
+  auto done = shutdown_manager->Start();
+  auto response = uut->Pull(3).get();
+  ASSERT_THAT(response.status(), StatusIs(StatusCode::kOk));
+  EXPECT_THAT(acks(*response), ElementsAre("ack-0-0", "ack-0-1", "ack-0-2"));
+  ASSERT_EQ(1, timers.size());
+
+  // Ack one of the messages and then fire the timer, the expectations set above
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_lease_management_test.cc:131]: s/, the/. The/

Comment (coryan) [reply google/cloud/pubsub/internal/subscription_lease_management_test.cc]: Done.

File: google/cloud/pubsub/internal/subscription_lease_management_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_lease_management_test.cc]: Done, but I am not sure it is a win... I am happy either way.

File: google/cloud/pubsub/internal/subscription_batch_source.h (hunk 4 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_batch_source.h]: Fixed.

File: google/cloud/pubsub/internal/subscription_batch_source.h (hunk 3 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_batch_source.h]: PTAL

File: google/cloud/pubsub/internal/subscription_batch_source.h (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_batch_source.h]: Fixed.

File: google/cloud/pubsub/internal/subscription_batch_source.h (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_batch_source.h]: Fixed.

File: google/cloud/pubsub/internal/subscription_lease_management_test.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_lease_management_test.cc:46]: Undone. MSVC did not like the 64-bit integer -> double conversions.

File: google/cloud/pubsub/internal/subscription_batch_source.h [hunk 7]
```diff
@@ -59,10 +59,10 @@ class SubscriptionBatchSource {
   /**
    * Negative acknowledgement for message associated with @p ack_id.
    *
-   * The application has not able to handle this message, the service should
-   * re-deliver the message (subject to the topic and subscription
-   * configuration). The @p size parameter should be the original message size
-   * estimate, it may be used by the message source to flow control large
+   * The application has not able to handle this message. Nacking a message
+   * allows the service re-deliver it, subject to the topic and subscription
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_batch_source.h:63]: s/re-deliver/to re-deliver/

File: google/cloud/pubsub/internal/subscription_lease_management_test.cc (hunk 5 above)
Comment (devbww) [reply google/cloud/pubsub/internal/subscription_lease_management_test.cc:46]: I'm not sure why `double` needs enter the picture, but OK.

[The win, I'd say, is to never be dealing with times/durations as raw integers, except right at an interface boundary where it is forced upon you.]

File: google/cloud/pubsub/internal/subscription_batch_source.h (hunk 7 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_batch_source.h]: Done.

**Pull request merged**